packages/core-transaction-pool/src/query.ts

```diff
@@ -0,0 +1,156 @@
+    private readonly iter: Iterator<Interfaces.ITransaction>;
+    private readonly predicate: Contracts.TransactionPool.Predicate | undefined;
+
+    public constructor(iter: Iterator<Interfaces.ITransaction>, predicate?: Contracts.TransactionPool.Predicate) {
+        this.iter = iter;
+        this.predicate = predicate;
+    }
+
+    public next(value?: any): IteratorResult<Interfaces.ITransaction> {
+        for (let r = this.iter.next(); !r.done; r = this.iter.next()) {
+            if (!this.predicate || this.predicate(r.value)) {
+                return r;
+            }
+        }
+
+        return { done: true, value: undefined };
+    }
+}
+
+class QueryIterable implements Contracts.TransactionPool.QueryIterable {
+    private readonly transactions: Iterable<Interfaces.ITransaction>;
+    private readonly predicate: Contracts.TransactionPool.Predicate | undefined;
+
+    public constructor(
+        transactions: Iterable<Interfaces.ITransaction>,
+        predicate?: Contracts.TransactionPool.Predicate,
+    ) {
+        this.transactions = transactions;
+        this.predicate = predicate;
+    }
+
+    public [Symbol.iterator](): Iterator<Interfaces.ITransaction> {
+        return new QueryIterator(this.transactions[Symbol.iterator](), this.predicate);
+    }
+
+    public whenPredicate(predicate: Contracts.TransactionPool.Predicate): Contracts.TransactionPool.QueryIterable {
+        return new QueryIterable(this, predicate);
+    }
+
+    public whenId(id: string): Contracts.TransactionPool.QueryIterable {
+        return this.whenPredicate(t => t.id === id);
+    }
+
+    public whenType(type: number): Contracts.TransactionPool.QueryIterable {
+        return this.whenPredicate(t => t.data.type === type);
+    }
+
+    public whenTypeGroup(typeGroup: number): Contracts.TransactionPool.QueryIterable {
+        return this.whenPredicate(t => t.data.typeGroup === typeGroup);
+    }
+
+    public whenVersion(version: number): Contracts.TransactionPool.QueryIterable {
+        return this.whenPredicate(t => t.data.version === version);
+    }
+
+    public whenInternalType(
+        internalType: Transactions.InternalTransactionType,
+    ): Contracts.TransactionPool.QueryIterable {
+        return this.whenPredicate(
+            t => Transactions.InternalTransactionType.from(t.data.type, t.data.typeGroup) === internalType,
+        );
+    }
+
+    public whenKind(transaction: Interfaces.ITransaction): Contracts.TransactionPool.QueryIterable {
+        return this.whenPredicate(
+            t =>
+                t.data.type === transaction.data.type &&
+                t.data.typeGroup === transaction.data.typeGroup &&
+                t.data.version === transaction.data.version,
+        );
+    }
+
+    public has(): boolean {
+        return this[Symbol.iterator]().next().done === false;
+    }
+
+    public first(): Interfaces.ITransaction {
+        const r = this[Symbol.iterator]().next();
```

Comment:
Avoid single character variable assignments unless they are used in a `for` loop for numerical increments. A brief name that makes it clear what `r` is saves another developer the time to figure out what is returned by `next()`.

Reply:
Renamed that into `result`